flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -464,7 +472,15 @@ public void onContainerStarted(ContainerId containerId, Map<String, ByteBuffer>
 
 	@Override
 	public void onContainerStatusReceived(ContainerId containerId, ContainerStatus containerStatus) {
-		// We are not interested in getting container status
+		// We fetch the status of the container from the previous attempts.
+		if (containerStatus.getState() == ContainerState.NEW) {
```

Comment:
ContainerStatus#getState() may only returns RUNNING(means it's on starting or started) or COMPLETE(means it has finished) in most hadoop versions, rare versions may contains NEW or SCHEDULED. So that I think this condition can be declared as not RUNNING here, and we should add a condition like  `if (containerStatus.getState() != ContainerState.COMPLETE)` for the resourceManagerClient#releaseAssignedContainer calling since there's no necessary to do that.

We also should handle this in onGetContainerStatusError method for containers that haven't been started via calling NMClient#startContainer yet by the last AM.

Last suggestion is to consider consistence when internal state may be updated inside, it can be handled by calling runAsyc(...).

Reply:
@TaoYang526
Are you suggesting that calling `NMClientAsync.getContainerStatusAsync` on a `NEW` container might result in `onGetContainerStatusError` on some Hadoop versions while `onContainerStatusReceived` on other versions?

If that is the case, I think we can have a common method handling releasing the container and removing it from the worker node map, which should be called in both `onContainerStatusReceived` with `if (containerStatus.getState() == ContaienrState.NEW)` and `onGetContainerStatusError`.

One more question, how do we now whether a container is `NEW` or there's some other problems in `onGetContainerStatusError`?

Reply:
> Are you suggesting that calling NMClientAsync.getContainerStatusAsync on a NEW container might result in onGetContainerStatusError on some Hadoop versions while onContainerStatusReceived on other versions?

No, they are coexisting in Hadoop, onContainerStatusReceived is for containers that already started by AM via calling NMClient#startContainers while onGetContainerStatusError is for containers that haven't been been started by AM.

> If that is the case, I think we can have a common method handling releasing the container and removing it from the worker node map

Yes, a common method is necessary.

> One more question, how do we now whether a container is NEW or there's some other problems in onGetContainerStatusError?

There maybe several causes for this handling, such as container is not found on NM or NM can't be connected, but they can be considered as a same problem: this container may be not useable for now since we can't get the status successfully,  I think we can just handle this as above no matter what the real cause is.

Reply:
I just have an offline discussion with Xintong and Taoyang.

The key point of this issue is: we need to release the container which is not started in the previous attempt or is not useable anymore. When we try to get the status of those containers:
- If it goes into `onContainerStatusReceived`: It means that the container has been started in the previous attempt. No matter what is the state, these containers will be released eventually by the existing `onStartContainerError` or `onContainersCompleted`. There is no need to handle them.
- If it goes into `onGetContainerStatusError`: It means that the container is not started in the previous attempt or other causes like NM lost. In such cases, the container is not useable and we need to release it and remove it from the `workerNodeMap`.

So, I will move the release logic here to onGetContainerStatusError.